### PR TITLE
Fix unhandled exception and magic constant.

### PR DESCRIPTION
**Pull request by quantumagi**
This PR:
- Fixes unhandled exceptions that lead to unwanted breaks when trying to debug the full node.
- Fixes a magic constant by declaring a variable to hold the constant
- Arrange some code according to cost to execute.

### CODE REVIEW

**Comment by fassadlr** [comment]
@quantumagi I've had a think about this and I'm not sure that we need a IP endpoint collection class? The AddNode and Connect collection of nodes in ConnectionManager will rarely have more than 5-10 addresses in them, so we aren't really "scoring" on performance here.

The collection of addresses in PeerAddressManager is where we can possibly benefit from such an implementation as there can be 10 000+ items in there. However, that collection is a thread-safe dictionary and already looks up addresses by a hashed version of IPEndpoint, which is quite performant as is.

What do you think?

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs [hunk 1]
```diff
@@ -89,34 +91,34 @@ public override PeerAddress FindPeerToConnectTo()
                     continue;
                 }
 
+                // If the peer is already connected just continue.
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: +1 for moving this logic earlier.

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs [hunk 2]
```diff
@@ -13,6 +13,8 @@ namespace Stratis.Bitcoin.P2P
     /// </summary>
     public sealed class PeerConnectorDiscovery : PeerConnector
     {
+        private const int MaximumPeerSelectionAttempts = 50;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: This actually fixes issue #955 . Can you change this to 5?

Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: Done.

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: Thanks.

**Comment by quantumagi** [comment]
@fassadlr, regarding the IPEndPointSet. If we are looping only 5 times (assuming this will remain unchanged) over 5 to 10 nodes we would still be converting (to IPV6) and scanning 25 to 50 addresses unnecessarily every few seconds. The class is only a few lines of code and is reducing complexity rather than adding to it.

**Comment by noescape00** [comment]
>Fixes unhandled exceptions that lead to unwanted breaks when trying to debug the full node.
@quantumagi this is not a problem with code, it's about the VS debugger. Please take a look at [this issue.](https://github.com/stratisproject/StratisBitcoinFullNode/issues/928)

**Comment by quantumagi** [comment]
@noescape00, at least this provides a convenient work-around. Let's go with this solution.

**Comment by noescape00** [comment]
@quantumagi Ok, I don't mind. Just wasn't sure if we should implement it before most of the people voted on this in the issue.

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs [hunk 3]
```diff
@@ -172,13 +172,29 @@ public void Listen(int maxIncoming = 20)
             {
                 while (!this.serverCancel.IsCancellationRequested)
                 {
+                    // Used to record any errors occurrring in the thread pool task.
+                    Exception err = null;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs]: Rename to "error" "or connectionException"

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs [hunk 4]
```diff
@@ -172,13 +172,29 @@ public void Listen(int maxIncoming = 20)
             {
                 while (!this.serverCancel.IsCancellationRequested)
                 {
+                    // Used to record any errors occurrring in the thread pool task.
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs]: 3 r's ;)

File: src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 ﻿using System.Collections.Generic;
 using System.Net;
+using Stratis.Bitcoin.Connection;
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs]: why this?

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs [hunk 6]
```diff
@@ -172,13 +172,29 @@ public void Listen(int maxIncoming = 20)
...
+                        catch (Exception e)
+                        {
+                            // Record the error.
+                            err = e;
+                            return null;                            
+                        }
                     }).ConfigureAwait(false);
 
+                    // Raise the error.
+                    if (err != null)
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs]: pls be consistent with the previous code, either throw in both on new line or same line

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs [hunk 7]
```diff
@@ -89,11 +92,19 @@ public override PeerAddress FindPeerToConnectTo()
...
+                    peerSelectionFailed++;
+                    continue;
+                }
+
                 // If the peer exists in the -addnode collection don't 
                 // try and connect to it.
                 var peerExistsInAddNode = this.NodeSettings.ConnectionManager.AddNode.Any(p => p.MapToIpv6().Match(peer.NetworkAddress.Endpoint));
                 if (peerExistsInAddNode)
-                {
+                {
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: delete extra spaces

Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: Done

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs (hunk 6 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs]: Done

File: src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs (hunk 5 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs]: Removed. It was left over after some optimization was reverted.

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs]: Done

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs]: Fixed.

**Comment by dangershony** [comment]
Hi guys this is really ugly code 🦄 
We should make an overload on Task that takes a cancellation.

**Comment by Aprogiena** [comment]
it is ugly yes, I'd go with Task<Exception> rather than using local variable outside of the scope, but ... well feel free to come with something better if you have time

**Pull request merged**